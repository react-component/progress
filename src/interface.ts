import * as React from 'react';

export interface ProgressProps {
  strokeWidth?: number;
  trailWidth?: number;
  className?: string;
  percent?: number | number[];
  strokeColor?: StrokeColorType;
  trailColor?: string;
  strokeLinecap?: StrokeLinecapType;
  prefixCls?: string;
  style?: React.CSSProperties;
  gapDegree?: number;
  gapPosition?: GapPositionType;
  transition?: string;
}

export type StrokeColorType = string | string[] | object;

export type GapPositionType = 'top' | 'right' | 'bottom' | 'left';

export type StrokeLinecapType = 'round' | 'butt' | 'square';
