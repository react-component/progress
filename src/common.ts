import { useRef, useEffect } from 'react';
import { ProgressProps } from './interface';

export const defaultProps: Partial<ProgressProps> = {
  className: '',
  percent: 0,
  prefixCls: 'rc-progress',
  strokeColor: '#2db7f5',
  strokeLinecap: 'round',
  strokeWidth: 1,
  style: {},
  trailColor: '#D9D9D9',
  trailWidth: 1,
};

export const useTransitionDuration = (percentList: number[]) => {
  // eslint-disable-next-line react-hooks/rules-of-hooks
  const paths = percentList.map(() => useRef());
  const prevTimeStamp = useRef(null);
  useEffect(() => {
    const now = Date.now();
    let updated = false;

    Object.keys(paths).forEach(key => {
      const path = paths[key].current;
      if (!path) {
        return;
      }
      updated = true;
      const pathStyle = path.style;
      pathStyle.transitionDuration = '.3s, .3s, .3s, .06s';

      if (prevTimeStamp.current && now - prevTimeStamp.current < 100) {
        pathStyle.transitionDuration = '0s, 0s';
      }
    });

    if (updated) {
      prevTimeStamp.current = Date.now();
    }
  });

  return [paths];
};
