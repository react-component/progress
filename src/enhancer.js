const enhancer = WrappedComponent => class Progress extends WrappedComponent {
  componentDidUpdate() {
    const now = Date.now();
    let updated = false;

    Object.keys(this.paths).forEach((key) => {
      const path = this.paths[key];

      if (!path) {
        return;
      }

      updated = true;
      const pathStyle = path.style;
      pathStyle.transitionDuration = '.3s, .3s, .3s, .06s';

      if (this.prevTimeStamp && now - this.prevTimeStamp < 100) {
        pathStyle.transitionDuration = '0s, 0s';
      }
    });

    if (updated) {
      this.prevTimeStamp = Date.now();
    }
  }

  render() {
    return super.render();
  }
};

export default enhancer;
