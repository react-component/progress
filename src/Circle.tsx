import * as React from 'react';
import classNames from 'classnames';
import { useTransitionDuration, defaultProps } from './common';
import type { ProgressProps } from './interface';
import useId from './hooks/useId';

function stripPercentToNumber(percent: string) {
  return +percent.replace('%', '');
}

function toArray<T>(value: T | T[]): T[] {
  const mergedValue = value ?? [];
  return Array.isArray(mergedValue) ? mergedValue : [mergedValue];
}

const VIEW_BOX_SIZE = 100;

const getCircleStyle = (
  radius: number,
  offset: number,
  percent: number,
  strokeColor: string | Record<string, string>,
  gapDegree = 0,
  gapPosition: ProgressProps['gapPosition'],
  strokeLinecap: ProgressProps['strokeLinecap'],
  strokeWidth,
) => {
  const rotateDeg = gapDegree > 0 ? 90 + gapDegree / 2 : -90;
  const perimeter = Math.PI * 2 * radius;
  const perimeterWithoutGap = perimeter * ((360 - gapDegree) / 360);
  const offsetDeg = (offset / 100) * 360 * ((360 - gapDegree) / 360);

  const positionDeg = {
    bottom: 0,
    top: 180,
    left: 90,
    right: -90,
  }[gapPosition];

  let strokeDashoffset = ((100 - percent) / 100) * perimeterWithoutGap;
  // Fix percent accuracy when strokeLinecap is round
  // https://github.com/ant-design/ant-design/issues/35009
  if (strokeLinecap === 'round' && percent !== 100) {
    strokeDashoffset += strokeWidth / 2;
    // when percent is small enough (<= 1%), keep smallest value to avoid it's disapperance
    if (strokeDashoffset >= perimeterWithoutGap) {
      strokeDashoffset = perimeterWithoutGap - 0.01;
    }
  }

  return {
    stroke: typeof strokeColor === 'string' ? strokeColor : undefined,
    strokeDasharray: `${perimeterWithoutGap}px ${perimeter}`,
    strokeDashoffset,
    transform: `rotate(${rotateDeg + offsetDeg + positionDeg}deg)`,
    transformOrigin: '50% 50%',
    transition:
      'stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s',
    fillOpacity: 0,
  };
};

const Circle: React.FC<ProgressProps> = ({
  id,
  prefixCls,
  strokeWidth,
  trailWidth,
  gapDegree,
  gapPosition,
  trailColor,
  strokeLinecap,
  style,
  className,
  strokeColor,
  percent,
  ...restProps
}) => {
  const mergedId = useId(id);
  const gradientId = `${mergedId}-gradient`;
  const radius = VIEW_BOX_SIZE / 2 - strokeWidth / 2;

  const circleStyle = getCircleStyle(
    radius,
    0,
    100,
    trailColor,
    gapDegree,
    gapPosition,
    strokeLinecap,
    strokeWidth,
  );
  const percentList = toArray(percent);
  const strokeColorList = toArray(strokeColor);
  const gradient = strokeColorList.find((color) => color && typeof color === 'object');

  const [paths] = useTransitionDuration(percentList);

  const getStokeList = () => {
    let stackPtg = 0;
    return percentList
      .map((ptg, index) => {
        const color = strokeColorList[index] || strokeColorList[strokeColorList.length - 1];
        const stroke = color && typeof color === 'object' ? `url(#${gradientId})` : undefined;
        const circleStyleForStack = getCircleStyle(
          radius,
          stackPtg,
          ptg,
          color,
          gapDegree,
          gapPosition,
          strokeLinecap,
          strokeWidth,
        );
        stackPtg += ptg;
        return (
          <circle
            key={index}
            className={`${prefixCls}-circle-path`}
            r={radius}
            cx={VIEW_BOX_SIZE / 2}
            cy={VIEW_BOX_SIZE / 2}
            stroke={stroke}
            strokeLinecap={strokeLinecap}
            strokeWidth={strokeWidth}
            opacity={ptg === 0 ? 0 : 1}
            style={circleStyleForStack}
            ref={paths[index]}
          />
        );
      })
      .reverse();
  };

  return (
    <svg
      className={classNames(`${prefixCls}-circle`, className)}
      viewBox={`0 0 ${VIEW_BOX_SIZE} ${VIEW_BOX_SIZE}`}
      style={style}
      id={id}
      {...restProps}
    >
      {gradient && (
        <defs>
          <linearGradient id={gradientId} x1="100%" y1="0%" x2="0%" y2="0%">
            {Object.keys(gradient)
              .sort((a, b) => stripPercentToNumber(a) - stripPercentToNumber(b))
              .map((key, index) => (
                <stop key={index} offset={key} stopColor={gradient[key]} />
              ))}
          </linearGradient>
        </defs>
      )}
      <circle
        className={`${prefixCls}-circle-trail`}
        r={radius}
        cx={VIEW_BOX_SIZE / 2}
        cy={VIEW_BOX_SIZE / 2}
        stroke={trailColor}
        strokeLinecap={strokeLinecap}
        strokeWidth={trailWidth || strokeWidth}
        style={circleStyle}
      />
      {getStokeList()}
    </svg>
  );
};

Circle.defaultProps = defaultProps;

Circle.displayName = 'Circle';

export default Circle;
