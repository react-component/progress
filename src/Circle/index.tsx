import * as React from 'react';
import classNames from 'classnames';
import { defaultProps, useTransitionDuration } from '../common';
import type { ProgressProps } from '../interface';
import useId from '../hooks/useId';
import PtgCircle from './PtgCircle';
import { VIEW_BOX_SIZE, getCircleStyle } from './util';

function toArray<T>(value: T | T[]): T[] {
  const mergedValue = value ?? [];
  return Array.isArray(mergedValue) ? mergedValue : [mergedValue];
}

const Circle: React.FC<ProgressProps> = (props) => {
  const {
    id,
    prefixCls,
    steps,
    strokeWidth,
    trailWidth,
    gapDegree = 0,
    gapPosition,
    trailColor,
    strokeLinecap,
    style,
    className,
    strokeColor,
    percent,
    ...restProps
  } = {
    ...defaultProps,
    ...props,
  };

  const halfSize = VIEW_BOX_SIZE / 2;

  const mergedId = useId(id);
  const gradientId = `${mergedId}-gradient`;
  const radius = halfSize - strokeWidth / 2;
  const perimeter = Math.PI * 2 * radius;
  const rotateDeg = gapDegree > 0 ? 90 + gapDegree / 2 : -90;
  const perimeterWithoutGap = perimeter * ((360 - gapDegree) / 360);
  const { count: stepCount, gap: stepGap } =
    typeof steps === 'object' ? steps : { count: steps, gap: 2 };

  const percentList = toArray(percent);
  const strokeColorList = toArray(strokeColor);
  const gradient = strokeColorList.find((color) => color && typeof color === 'object') as Record<
    string,
    string
  >;
  const isConicGradient = gradient && typeof gradient === 'object';
  const mergedStrokeLinecap = isConicGradient ? 'butt' : strokeLinecap;

  const circleStyle = getCircleStyle(
    perimeter,
    perimeterWithoutGap,
    0,
    100,
    rotateDeg,
    gapDegree,
    gapPosition,
    trailColor,
    mergedStrokeLinecap,
    strokeWidth,
  );

  const paths = useTransitionDuration();

  const getStokeList = () => {
    let stackPtg = 0;
    return percentList
      .map<React.ReactNode>((ptg, index) => {
        const color = strokeColorList[index] || strokeColorList[strokeColorList.length - 1];
        const circleStyleForStack = getCircleStyle(
          perimeter,
          perimeterWithoutGap,
          stackPtg,
          ptg,
          rotateDeg,
          gapDegree,
          gapPosition,
          color,
          mergedStrokeLinecap,
          strokeWidth,
        );
        stackPtg += ptg;

        return (
          <PtgCircle
            key={index}
            color={color}
            ptg={ptg}
            radius={radius}
            prefixCls={prefixCls}
            gradientId={gradientId}
            style={circleStyleForStack}
            strokeLinecap={mergedStrokeLinecap}
            strokeWidth={strokeWidth}
            gapDegree={gapDegree}
            ref={(elem) => {
              // https://reactjs.org/docs/refs-and-the-dom.html#callback-refs
              // React will call the ref callback with the DOM element when the component mounts,
              // and call it with `null` when it unmounts.
              // Refs are guaranteed to be up-to-date before componentDidMount or componentDidUpdate fires.

              paths[index] = elem;
            }}
            size={VIEW_BOX_SIZE}
          />
        );
      })
      .reverse();
  };

  const getStepStokeList = () => {
    // only show the first percent when pass steps
    const current = Math.round(stepCount * (percentList[0] / 100));
    const stepPtg = 100 / stepCount;

    let stackPtg = 0;
    return new Array(stepCount).fill(null).map<React.ReactNode>((_, index) => {
      const color = index <= current - 1 ? strokeColorList[0] : trailColor;
      const stroke = color && typeof color === 'object' ? `url(#${gradientId})` : undefined;
      const circleStyleForStack = getCircleStyle(
        perimeter,
        perimeterWithoutGap,
        stackPtg,
        stepPtg,
        rotateDeg,
        gapDegree,
        gapPosition,
        color,
        'butt',
        strokeWidth,
        stepGap,
      );
      stackPtg +=
        ((perimeterWithoutGap - (circleStyleForStack.strokeDashoffset as number) + stepGap) * 100) /
        perimeterWithoutGap;

      return (
        <circle
          key={index}
          className={`${prefixCls}-circle-path`}
          r={radius}
          cx={halfSize}
          cy={halfSize}
          stroke={stroke}
          strokeWidth={strokeWidth}
          opacity={1}
          style={circleStyleForStack}
          ref={(elem) => {
            paths[index] = elem;
          }}
        />
      );
    });
  };

  return (
    <svg
      className={classNames(`${prefixCls}-circle`, className)}
      viewBox={`0 0 ${VIEW_BOX_SIZE} ${VIEW_BOX_SIZE}`}
      style={style}
      id={id}
      role="presentation"
      {...restProps}
    >
      {!stepCount && (
        <circle
          className={`${prefixCls}-circle-trail`}
          r={radius}
          cx={halfSize}
          cy={halfSize}
          stroke={trailColor}
          strokeLinecap={mergedStrokeLinecap}
          strokeWidth={trailWidth || strokeWidth}
          style={circleStyle}
        />
      )}
      {stepCount ? getStepStokeList() : getStokeList()}
    </svg>
  );
};

if (process.env.NODE_ENV !== 'production') {
  Circle.displayName = 'Circle';
}

export default Circle;
