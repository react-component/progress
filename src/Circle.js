/* eslint react/prop-types: 0 */
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import enhancer from './enhancer';
import { propTypes, defaultProps } from './types';

class Circle extends Component {
  getPathStyles() {
    const { percent, strokeWidth, gapDegree = 0, gapPosition } = this.props;
    const radius = 50 - (strokeWidth / 2);
    let beginPositionX = 0;
    let beginPositionY = -radius;
    let endPositionX = 0;
    let endPositionY = -2 * radius;
    switch (gapPosition) {
      case 'left':
        beginPositionX = -radius;
        beginPositionY = 0;
        endPositionX = 2 * radius;
        endPositionY = 0;
        break;
      case 'right':
        beginPositionX = radius;
        beginPositionY = 0;
        endPositionX = -2 * radius;
        endPositionY = 0;
        break;
      case 'bottom':
        beginPositionY = radius;
        endPositionY = 2 * radius;
        break;
      default:
    }
    const pathString = `M 50,50 m ${beginPositionX},${beginPositionY}
     a ${radius},${radius} 0 1 1 ${endPositionX},${-endPositionY}
     a ${radius},${radius} 0 1 1 ${-endPositionX},${endPositionY}`;
    const len = Math.PI * 2 * radius;
    const trailPathStyle = {
      strokeDasharray: `${len - gapDegree}px ${len}px`,
      strokeDashoffset: `-${gapDegree / 2}px`,
      transition: 'stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s',
    };
    const strokePathStyle = {
      strokeDasharray: `${(percent / 100) * (len - gapDegree)}px ${len}px`,
      strokeDashoffset: `-${gapDegree / 2}px`,
      transition: 'stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s', // eslint-disable-line
    };
    return { pathString, trailPathStyle, strokePathStyle };
  }

  render() {
    const {
      prefixCls, strokeWidth, trailWidth, strokeColor, percent,
      trailColor, strokeLinecap, trailLinecap, style, className, ...restProps,
    } = this.props;
    const { pathString, trailPathStyle, strokePathStyle } = this.getPathStyles();
    delete restProps.percent;
    delete restProps.gapDegree;
    delete restProps.gapPosition;
    return (
      <svg
        className={`${prefixCls}-circle ${className}`}
        viewBox="0 0 100 100"
        style={style}
        {...restProps}
      >
        <path
          className={`${prefixCls}-circle-trail`}
          d={pathString}
          stroke={trailColor}
          strokeWidth={trailWidth || strokeWidth}
          strokeLinecap={trailLinecap}
          fillOpacity="0"
          style={trailPathStyle}
        />
        <path
          className={`${prefixCls}-circle-path`}
          d={pathString}
          strokeLinecap={strokeLinecap}
          stroke={strokeColor}
          strokeWidth={this.props.percent === 0 ? 0 : strokeWidth}
          fillOpacity="0"
          ref={(path) => { this.path = path; }}
          style={strokePathStyle}
        />
      </svg>
    );
  }
}

Circle.propTypes = {
  ...propTypes,
  gapPosition: PropTypes.oneOf(['top', 'bottom', 'left', 'right']),
};

Circle.defaultProps = {
  ...defaultProps,
  gapPosition: 'top',
};

export default enhancer(Circle);
