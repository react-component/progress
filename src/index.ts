import Line from './Line';
import Circle from './Circle';

export type { ProgressProps } from './interface';
export { Line, Circle };
export default {
  Line,
  Circle,
};
