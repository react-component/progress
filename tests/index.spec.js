/* eslint-disable react/no-render-return-value */
// eslint-disable-next-line max-classes-per-file
import React from 'react';
import { mount } from 'enzyme';
import { Line, Circle } from '../src';

describe('Progress', () => {
  describe('Line', () => {
    it('change with animation', () => {
      class Demo extends React.Component {
        state = {
          percent: '0',
        };

        render() {
          const { percent } = this.state;
          return <Line percent={percent} strokeWidth="1" />;
        }
      }

      const line = mount(<Demo />);
      expect(line.state().percent).toBe('0');
      line.setState({ percent: '30' });
      expect(line.state().percent).toBe('30');
      line.unmount();
    });
  });

  describe('Diff Line', () => {
    const wrapper = mount(
      <>
        <Line percent={20} strokeLinecap="butt" />
        <br />
        <Line percent={20} strokeLinecap="round" />
        <br />
        <Line percent={20} strokeLinecap="square" />
      </>,
    );
    expect(wrapper).toMatchSnapshot();
  });

  describe('Circle', () => {
    it('change with animation', () => {
      class Demo extends React.Component {
        state = {
          percent: '0',
        };

        render() {
          const { percent } = this.state;
          return <Circle percent={percent} strokeWidth="1" />;
        }
      }
      const circle = mount(<Demo />);
      expect(circle.state().percent).toBe('0');
      circle.setState({ percent: '30' });
      expect(circle.state().percent).toBe('30');
      circle.unmount();
    });

    it('should gradient works and circles have different gradient IDs', () => {
      const wrapper = mount(
        <div>
          <Circle
            percent={90}
            strokeWidth="6"
            strokeLinecap="round"
            strokeColor={{
              '0%': '#108ee9',
              '100%': '#87d068',
            }}
          />
          <Circle
            percent={90}
            strokeWidth="6"
            strokeLinecap="round"
            strokeColor={{
              '0%': '#108ee9',
              '100%': '#87d068',
            }}
          />
        </div>,
      );

      expect(wrapper.render()).toMatchSnapshot();
    });

    it('should show right gapPosition', () => {
      const wrapper = mount(
        <>
          <Circle
            percent={30}
            gapDegree={70}
            gapPosition="top"
            strokeWidth="6"
            strokeLinecap="square"
          />
          <Circle
            percent={30}
            gapDegree={70}
            gapPosition="bottom"
            strokeWidth="6"
            strokeLinecap="square"
          />
          <Circle
            percent={30}
            gapDegree={70}
            gapPosition="left"
            strokeWidth="6"
            strokeLinecap="square"
          />
          <Circle
            percent={30}
            gapDegree={70}
            gapPosition="right"
            strokeWidth="6"
            strokeLinecap="square"
          />
          <Circle
            percent={30}
            gapDegree={70}
            gapPosition="top"
            strokeWidth="6"
            strokeLinecap="round"
          />
          <Circle percent={30} gapDegree={70} gapPosition="top" strokeWidth="6" />
        </>,
      );

      expect(wrapper).toMatchSnapshot();
    });

    // https://github.com/ant-design/ant-design/issues/30552
    it('should change strokeColor between gradient and color string correctly', () => {
      const gradientColor = {
        '0%': '#108ee9',
        '100%': '#87d068',
      };
      const wrapper = mount(<Circle strokeColor={gradientColor} />);
      expect(wrapper.find('.rc-progress-circle-path').getDOMNode().style.cssText).not.toContain(
        'stroke:',
      );
      wrapper.setProps({ strokeColor: '#eeeeee' });
      expect(wrapper.find('.rc-progress-circle-path').getDOMNode().style.cssText).toContain(
        'stroke: #eeeeee',
      );
      wrapper.setProps({ strokeColor: gradientColor });
      expect(wrapper.find('.rc-progress-circle-path').getDOMNode().style.cssText).not.toContain(
        'stroke:',
      );
    });
  });
});
