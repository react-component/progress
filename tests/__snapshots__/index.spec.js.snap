// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` 1`] = `
<div>
  <svg
    class="rc-progress-line"
    preserveAspectRatio="none"
    viewBox="0 0 100 1"
  >
    <path
      class="rc-progress-line-trail"
      d="M 0,0.5
         L 100,0.5"
      fill-opacity="0"
      stroke="#D9D9D9"
      stroke-linecap="butt"
      stroke-width="1"
    />
    <path
      class="rc-progress-line-path"
      d="M 0,0.5
         L 100,0.5"
      fill-opacity="0"
      stroke="#2db7f5"
      stroke-linecap="butt"
      stroke-width="1"
      style="stroke-dasharray: 20px, 100px; stroke-dashoffset: -0px; transition: stroke-dashoffset 0.3s ease 0s, stroke-dasharray .3s ease 0s, stroke 0.3s linear; transition-duration: .3s, .3s, .3s, .06s;"
    />
  </svg>
  <br />
  <svg
    class="rc-progress-line"
    preserveAspectRatio="none"
    viewBox="0 0 100 1"
  >
    <path
      class="rc-progress-line-trail"
      d="M 0.5,0.5
         L 99.5,0.5"
      fill-opacity="0"
      stroke="#D9D9D9"
      stroke-linecap="round"
      stroke-width="1"
    />
    <path
      class="rc-progress-line-path"
      d="M 0.5,0.5
         L 99.5,0.5"
      fill-opacity="0"
      stroke="#2db7f5"
      stroke-linecap="round"
      stroke-width="1"
      style="stroke-dasharray: 19.8px, 100px; stroke-dashoffset: -0px; transition: stroke-dashoffset 0.3s ease 0s, stroke-dasharray .3s ease 0s, stroke 0.3s linear; transition-duration: .3s, .3s, .3s, .06s;"
    />
  </svg>
  <br />
  <svg
    class="rc-progress-line"
    preserveAspectRatio="none"
    viewBox="0 0 100 1"
  >
    <path
      class="rc-progress-line-trail"
      d="M 0,0.5
         L 100,0.5"
      fill-opacity="0"
      stroke="#D9D9D9"
      stroke-linecap="square"
      stroke-width="1"
    />
    <path
      class="rc-progress-line-path"
      d="M 0,0.5
         L 100,0.5"
      fill-opacity="0"
      stroke="#2db7f5"
      stroke-linecap="square"
      stroke-width="1"
      style="stroke-dasharray: 19.9px, 100px; stroke-dashoffset: -0px; transition: stroke-dashoffset 0.3s ease 0s, stroke-dasharray .3s ease 0s, stroke 0.3s linear; transition-duration: .3s, .3s, .3s, .06s;"
    />
  </svg>
</div>
`;

exports[`Progress Circle should gradient works and circles have different gradient IDs 1`] = `
<div>
  <svg
    class="rc-progress-circle"
    viewBox="0 0 100 100"
  >
    <defs>
      <linearGradient
        id="rc_progress_TEST_OR_SSR-gradient"
        x1="100%"
        x2="0%"
        y1="0%"
        y2="0%"
      >
        <stop
          offset="0%"
          stop-color="#108ee9"
        />
        <stop
          offset="100%"
          stop-color="#87d068"
        />
      </linearGradient>
    </defs>
    <circle
      class="rc-progress-circle-trail"
      cx="50"
      cy="50"
      r="47"
      stroke="#D9D9D9"
      stroke-linecap="round"
      stroke-width="1"
      style="stroke: #D9D9D9; stroke-dasharray: 295.3097094374406px 295.3097094374406; stroke-dashoffset: 0; transform: rotate(-90deg); transform-origin: 50% 50%; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s; fill-opacity: 0;"
    />
    <circle
      class="rc-progress-circle-path"
      cx="50"
      cy="50"
      opacity="1"
      r="47"
      stroke="url(#rc_progress_TEST_OR_SSR-gradient)"
      stroke-linecap="round"
      stroke-width="6"
      style="stroke-dasharray: 295.3097094374406px 295.3097094374406; stroke-dashoffset: 32.53097094374406; transform: rotate(-90deg); transform-origin: 50% 50%; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s; fill-opacity: 0; transition-duration: 0s, 0s;"
    />
  </svg>
  <svg
    class="rc-progress-circle"
    viewBox="0 0 100 100"
  >
    <defs>
      <linearGradient
        id="rc_progress_TEST_OR_SSR-gradient"
        x1="100%"
        x2="0%"
        y1="0%"
        y2="0%"
      >
        <stop
          offset="0%"
          stop-color="#108ee9"
        />
        <stop
          offset="100%"
          stop-color="#87d068"
        />
      </linearGradient>
    </defs>
    <circle
      class="rc-progress-circle-trail"
      cx="50"
      cy="50"
      r="47"
      stroke="#D9D9D9"
      stroke-linecap="round"
      stroke-width="1"
      style="stroke: #D9D9D9; stroke-dasharray: 295.3097094374406px 295.3097094374406; stroke-dashoffset: 0; transform: rotate(-90deg); transform-origin: 50% 50%; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s; fill-opacity: 0;"
    />
    <circle
      class="rc-progress-circle-path"
      cx="50"
      cy="50"
      opacity="1"
      r="47"
      stroke="url(#rc_progress_TEST_OR_SSR-gradient)"
      stroke-linecap="round"
      stroke-width="6"
      style="stroke-dasharray: 295.3097094374406px 295.3097094374406; stroke-dashoffset: 32.53097094374406; transform: rotate(-90deg); transform-origin: 50% 50%; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s; fill-opacity: 0; transition-duration: 0s, 0s;"
    />
  </svg>
</div>
`;

exports[`Progress Circle should show right gapPosition 1`] = `
<div>
  <svg
    class="rc-progress-circle"
    viewBox="0 0 100 100"
  >
    <circle
      class="rc-progress-circle-trail"
      cx="50"
      cy="50"
      r="47"
      stroke="#D9D9D9"
      stroke-linecap="square"
      stroke-width="1"
      style="stroke: #D9D9D9; stroke-dasharray: 237.88837704682715px 295.3097094374406; stroke-dashoffset: 0; transform: rotate(305deg); transform-origin: 50% 50%; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s; fill-opacity: 0;"
    />
    <circle
      class="rc-progress-circle-path"
      cx="50"
      cy="50"
      opacity="1"
      r="47"
      stroke-linecap="square"
      stroke-width="6"
      style="stroke: #2db7f5; stroke-dasharray: 237.88837704682715px 295.3097094374406; stroke-dashoffset: 166.52186393277898; transform: rotate(305deg); transform-origin: 50% 50%; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s; fill-opacity: 0; transition-duration: 0s, 0s;"
    />
  </svg>
  <svg
    class="rc-progress-circle"
    viewBox="0 0 100 100"
  >
    <circle
      class="rc-progress-circle-trail"
      cx="50"
      cy="50"
      r="47"
      stroke="#D9D9D9"
      stroke-linecap="square"
      stroke-width="1"
      style="stroke: #D9D9D9; stroke-dasharray: 237.88837704682715px 295.3097094374406; stroke-dashoffset: 0; transform: rotate(125deg); transform-origin: 50% 50%; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s; fill-opacity: 0;"
    />
    <circle
      class="rc-progress-circle-path"
      cx="50"
      cy="50"
      opacity="1"
      r="47"
      stroke-linecap="square"
      stroke-width="6"
      style="stroke: #2db7f5; stroke-dasharray: 237.88837704682715px 295.3097094374406; stroke-dashoffset: 166.52186393277898; transform: rotate(125deg); transform-origin: 50% 50%; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s; fill-opacity: 0; transition-duration: 0s, 0s;"
    />
  </svg>
  <svg
    class="rc-progress-circle"
    viewBox="0 0 100 100"
  >
    <circle
      class="rc-progress-circle-trail"
      cx="50"
      cy="50"
      r="47"
      stroke="#D9D9D9"
      stroke-linecap="square"
      stroke-width="1"
      style="stroke: #D9D9D9; stroke-dasharray: 237.88837704682715px 295.3097094374406; stroke-dashoffset: 0; transform: rotate(215deg); transform-origin: 50% 50%; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s; fill-opacity: 0;"
    />
    <circle
      class="rc-progress-circle-path"
      cx="50"
      cy="50"
      opacity="1"
      r="47"
      stroke-linecap="square"
      stroke-width="6"
      style="stroke: #2db7f5; stroke-dasharray: 237.88837704682715px 295.3097094374406; stroke-dashoffset: 166.52186393277898; transform: rotate(215deg); transform-origin: 50% 50%; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s; fill-opacity: 0; transition-duration: 0s, 0s;"
    />
  </svg>
  <svg
    class="rc-progress-circle"
    viewBox="0 0 100 100"
  >
    <circle
      class="rc-progress-circle-trail"
      cx="50"
      cy="50"
      r="47"
      stroke="#D9D9D9"
      stroke-linecap="square"
      stroke-width="1"
      style="stroke: #D9D9D9; stroke-dasharray: 237.88837704682715px 295.3097094374406; stroke-dashoffset: 0; transform: rotate(35deg); transform-origin: 50% 50%; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s; fill-opacity: 0;"
    />
    <circle
      class="rc-progress-circle-path"
      cx="50"
      cy="50"
      opacity="1"
      r="47"
      stroke-linecap="square"
      stroke-width="6"
      style="stroke: #2db7f5; stroke-dasharray: 237.88837704682715px 295.3097094374406; stroke-dashoffset: 166.52186393277898; transform: rotate(35deg); transform-origin: 50% 50%; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s; fill-opacity: 0; transition-duration: 0s, 0s;"
    />
  </svg>
  <svg
    class="rc-progress-circle"
    viewBox="0 0 100 100"
  >
    <circle
      class="rc-progress-circle-trail"
      cx="50"
      cy="50"
      r="47"
      stroke="#D9D9D9"
      stroke-linecap="round"
      stroke-width="1"
      style="stroke: #D9D9D9; stroke-dasharray: 237.88837704682715px 295.3097094374406; stroke-dashoffset: 0; transform: rotate(305deg); transform-origin: 50% 50%; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s; fill-opacity: 0;"
    />
    <circle
      class="rc-progress-circle-path"
      cx="50"
      cy="50"
      opacity="1"
      r="47"
      stroke-linecap="round"
      stroke-width="6"
      style="stroke: #2db7f5; stroke-dasharray: 237.88837704682715px 295.3097094374406; stroke-dashoffset: 169.52186393277898; transform: rotate(305deg); transform-origin: 50% 50%; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s; fill-opacity: 0; transition-duration: 0s, 0s;"
    />
  </svg>
  <svg
    class="rc-progress-circle"
    viewBox="0 0 100 100"
  >
    <circle
      class="rc-progress-circle-trail"
      cx="50"
      cy="50"
      r="47"
      stroke="#D9D9D9"
      stroke-linecap="round"
      stroke-width="1"
      style="stroke: #D9D9D9; stroke-dasharray: 237.88837704682715px 295.3097094374406; stroke-dashoffset: 0; transform: rotate(305deg); transform-origin: 50% 50%; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s; fill-opacity: 0;"
    />
    <circle
      class="rc-progress-circle-path"
      cx="50"
      cy="50"
      opacity="1"
      r="47"
      stroke-linecap="round"
      stroke-width="6"
      style="stroke: #2db7f5; stroke-dasharray: 237.88837704682715px 295.3097094374406; stroke-dashoffset: 169.52186393277898; transform: rotate(305deg); transform-origin: 50% 50%; transition: stroke-dashoffset .3s ease 0s, stroke-dasharray .3s ease 0s, stroke .3s, stroke-width .06s ease .3s, opacity .3s ease 0s; fill-opacity: 0; transition-duration: 0s, 0s;"
    />
  </svg>
</div>
`;
